### DISCUSSION TITLE
Remove empty albums

### DISCUSSION THREAD

**Discussion by habi** [RhetTbull/osxphotos#1315] [Q&A] [answered]
From a botched merge of libraries with [PowerPhotos](https://fatcatsoftware.com/powerphotos/) a *long* time ago I have very large amount of duplicate albums in Photos, several hierarchies deep (see screenshot1).
Merging/deleting these manually, is a *big* chore, which I've put off for much too long.

A lot of these albums are empty, and also show up as such when I call `osxphotos albums` (see screenshot2).

Is there a way to use `osxphotos` to programmatically remove *all* empty albums in my library?

osxphotos 0.65.0
Photos Version 9.0 (610.3.110)
macOS 14.1.2 (23B92)

<img width="246" alt="Screenshot 2023-12-06 at 23 53 51" src="https://github.com/RhetTbull/osxphotos/assets/1651235/dfcadec1-c039-4159-85d8-1ee6f0dfb4b3">

<img width="193" alt="Screenshot 2023-12-06 at 23 55 38" src="https://github.com/RhetTbull/osxphotos/assets/1651235/77d1813f-46fe-4368-89b3-a419c415b72b">

**Comment by RhetTbull** [comment]
This isn't possible out of the box but via the PhotoScript library (another of my projects) it is possible to delete albums and folders. I think it would not be hard to write a script that osxphotos could run to do this. Let me think about it and get back to you.

**Reply by habi** [reply]
BTW: kind of embarrassing, I've asked the same question there a while ago: https://github.com/RhetTbull/PhotoScript/pull/26#issuecomment-945684882
Shows how long I've been procrastinating on fixing this :)

**Comment by RhetTbull** [comment, marked as answer]
@habi I wrote a [quick script](https://github.com/RhetTbull/osxphotos/blob/main/examples/delete_empty_albums.py) to do this that can be run with osxphotos (version 0.65.0 and later).  You can run the script directly from GitHub without downloading it first via:

` osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/delete_empty_albums.py --folders --dry-run`

It has only 2 options: `--folders`: prune empty folders and `--dry-run`: show what will be done without actually deleting anything.

There is one limitation: empty subfolders cannot be deleted (though these will be listed so you can manually delete them). This is due to a bug in Apple Photos that frustratingly has been there since macOS Catalina. I'm on Ventura and it's still not fixed (have not tested Sonoma).
